salt/netapi/rest_tornado/saltnado.py

```diff
@@ -925,64 +947,83 @@ def _disbatch_local(self, chunk):
         if self.application.opts['order_masters']:
             syndic_min_wait = tornado.gen.sleep(self.application.opts['syndic_wait'])
 
-        job_not_running = self.job_not_running(pub_data['jid'],
-                                               chunk['tgt'],
-                                               f_call['kwargs']['tgt_type'],
-                                               minions_remaining=minions_remaining
-                                               )
+        # To ensure job_not_running and all_return are terminated by each other, communicate using a future
+        is_finished = Future()
+        job_not_running_future = self.job_not_running(pub_data['jid'],
+                                                      chunk['tgt'],
+                                                      f_call['kwargs']['tgt_type'],
+                                                      is_finished,
+                                                      minions_remaining=list(minions_remaining),
+                                                      )
 
         # if we have a min_wait, do that
         if syndic_min_wait is not None:
             yield syndic_min_wait
-        # we are completed when either all minions return or the job isn't running anywhere
-        chunk_ret = yield self.all_returns(pub_data['jid'],
-                                           finish_futures=[job_not_running],
-                                           minions_remaining=minions_remaining,
-                                           )
 
-        raise tornado.gen.Return(chunk_ret)
+        all_return_future = self.all_returns(pub_data['jid'],
+                                             is_finished,
+                                             minions_remaining=list(minions_remaining),
+                                             )
+        yield job_not_running_future
+        raise tornado.gen.Return((yield all_return_future))
```

Comment:
Are the double parenthesis required here? Would this work:
```
raise tornado.gen.Return(yield all_return_future)
```

Reply:
Inner parentheses is part of yield expressions. so it is required. 
you can see [doc](https://docs.python.org/3/reference/simple_stmts.html#the-yield-statement)